scenario/state.py

```diff
@@ -140,6 +141,77 @@ def copy(self) -> "Self":
         return copy.deepcopy(self)
 
 
+@dataclasses.dataclass(frozen=True)
+class CloudCredential:
+    auth_type: str
+    """Authentication type."""
+
+    attributes: Dict[str, str] = dataclasses.field(default_factory=dict)
+    """A dictionary containing cloud credentials.
+
+    For example, for AWS, it contains `access-key` and `secret-key`;
+    for Azure, `application-id`, `application-password` and `subscription-id`
+    can be found here.
+    """
+
+    redacted: List[str] = dataclasses.field(default_factory=list)
+    """A list of redacted secrets."""
+
+    def _to_ops(self) -> ops.CloudCredential:
+        return ops.CloudCredential(
+            auth_type=self.auth_type,
+            attributes=self.attributes,
+            redacted=self.redacted,
+        )
+
+
+@dataclasses.dataclass(frozen=True)
+class CloudSpec:
+    type: str
+    """Type of the cloud."""
+
+    name: str = "localhost"
+    """Juju cloud name."""
+
+    region: Optional[str] = None
```

Comment:
Question for all: @benhoyt @tonyandrewmeyer @IronCore864 
do we feel that the whole cloud spec tree should be defaulted? 
As in, should the user be able to, out of the box, do:
```python
with scenario.Context(Foo).manager('start', State()) as mgr:
    assert mgr.charm.model.get_cloud_spec().endpoint
```
and get something meaningful?

I feel like that's a yes, and in that case we need to add default (factories) to all fields except perhaps credentials and ca certs.

WDYT?

Reply:
note that this means we'd also need to lift the consistency check and raise instead from mocking.py if the user tries to get the value.

Reply:
Default: it's hard to provide default values when you don't know what cloud the user is using. If we default to lxd/localhost but the user is using AWS, it makes no sense. We also don't know what the majority of the users are using, maybe AWS, GCP, Azure, etc. I prefer not to provide a default value since it totally depends on the user.

On the assertion, AFAIK the onlything meaningful is CloudCredential, according to existing charms that call the credential-get hook tool. All they ever need is CloudCredential in order to acces cloud APIs. I think we can change it to assert on CloudCredential instead of on the CloudSpec object, but I don't think that's necessary since CloudSpec contains CloudCredential.

Reply:
> Question for all: @benhoyt @tonyandrewmeyer @IronCore864 do we feel that the whole cloud spec tree should be defaulted? As in, should the user be able to, out of the box, do:
> 
> ```python
> with scenario.Context(Foo).manager('start', State()) as mgr:
>     assert mgr.charm.model.get_cloud_spec().endpoint
> ```
> 
> and get something meaningful?
> 
> I feel like that's a yes, and in that case we need to add default (factories) to all fields except perhaps credentials and ca certs.
> 
> WDYT?

Here's an [example charm that uses this functionality](https://github.com/charmed-kubernetes/charm-aws-integrator) (I don't know why Charmed Kubernetes is a bunch of forks of archived repos, but I assume that is unimportant).

It's currently [calling the hook tool directly](https://github.com/charmed-kubernetes/charm-aws-integrator/blob/26ad2b47f6c8f9cc231c24079cce9c59a1d40652/lib/charms/layer/aws.py#L45) but could now update to newer ops and use `get_cloud_spec` instead. This is an [example of a charm that uses that lib](https://github.com/charmed-kubernetes/charm-kubernetes-worker/blob/main/src/cloud_integration.py) - I can't find a simpler use of `credential-get` (e.g. one outside of a lib in a maintained charm), unfortunately.

If I'm testing that charm, then I think I'm writing a scenario something like:

```python
model = scenario.Model(type="lxd")
state_in = scenario.State(model=model)
ctx = scenario.Context(CharmName)
state_out = ctx.run("config-changed", state_in)
config_file = ...
config = configparser.ConfigParser()
config.read(config_file)
assert config_file.get("juju", "aws_access_key_id") == state_out.cloud_spec.credential.attributes["access-key"]
assert config_file.get("juju", "aws_secret_key_id") == state_out.cloud_spec.credential.attributes["secret-key"]
```

*(For that charm, I would have to do some mocking because it writes the file to `/root/` and we don't (currently) provide a mock filesystem for the charm that will work with the built-in `open`.)*

What I care about here is that after my event I have a config file generated in the expected format with the expected values. I probably would have another couple of tests that check that I get a config file generated in the expected format when there's no cloud spec (the code has alternative ways of getting the credentials if there's no `juju trust`).

Most of the time, I *don't* care, in a **unit** test, whether the provided values are meaningful. It seems like this would be the usual case - I don't care what the values are, just that they go where they are meant to. There might be some exceptions to this - maybe the event handler is doing something that makes use of the certificates somehow (not actually connecting to the cloud) and so they need to be a certificate (not a valid one, you couldn't put that in version control, but one that's valid). This is when you can provide other values, of course.

Unfortunately, it doesn't seem like we have a lot of examples of people using `credential-get` to really judge this, and I certainly don't have that experience myself. We could ask charmers but I'm sceptical that we'd get a lot of info.

I think it would be convenient to have default credentials, so you can do a Scenario like the above without putting them in yourself. The annoying part is that they are different based on the `type` so you'd have to have a default `allclouds` type or something that had dummy values for all possible attributes, which is not well defined. I feel like this gets too messy.

If I understand correctly, `CloudSpec.name` is an arbitrary string provided by the Juju admin when configuring the cloud. If that's right, then I think this should definitely be defaulted in the same way the model name is (there's another thread here somewhere where that's suggested).

That only leaves `type`. Given that it seems likely that you're going to have to provide some sort of extra values (the credentials) then I think putting in a `type` is ok.

I think if we see a lot of charmers start using `get_cloud_spec` then what we could do is have something like:

```python
class CloudSpec
    ...
    @classmethod
    def aws(cls):
        return cls(type="aws", credential=Credential(attributes={"secret-key": "secret", "access-key": "1234"})
    @classmethod
    def azure(cls):
        return ...
    ...
```

But I think it's fine to wait until there's demand for that.